File: backend/src/application/usecase/CreatorProductUseCase.go
package usecase

import (
	"github.com/marcosfrancomarinho/go-first-project/src/application/dto"
	"github.com/marcosfrancomarinho/go-first-project/src/domain/entities"
	"github.com/marcosfrancomarinho/go-first-project/src/domain/interfaces"
	"github.com/marcosfrancomarinho/go-first-project/src/domain/valuesobject"
)

type CreatorProductUseCase struct {
	creatorProduct interfaces.CreatorProduct
	geratorId      interfaces.GeratorID
}

func NewCreatorProductUseCase(
	creatorProduct interfaces.CreatorProduct,
	geratorId interfaces.GeratorID,
) *CreatorProductUseCase {
	return &CreatorProductUseCase{creatorProduct: creatorProduct, geratorId: geratorId}
}

func (c *CreatorProductUseCase) Create(payload *dto.RequestCreatorProductDTO) (*dto.ResponseCreatorProductDTO, error) {
	name, err := valuesobject.NewName(payload.Name)
	if err != nil {
		return nil, err
	}

	price, err := valuesobject.NewPrice(payload.Price)
	if err != nil {
		return nil, err
	}

	quantity, err := valuesobject.NewQuantity(payload.Quantity)
	if err != nil {
		return nil, err
	}

	id, err := c.geratorId.Generate()
	if err != nil {
		return nil, err
	}

	product := entities.NewProduct(id, name, price, quantity)

	if err := c.creatorProduct.Create(product); err != nil {
		return nil, err
	}
	
	return &dto.ResponseCreatorProductDTO{
		Id:      id.GetValue(),
		Message: "produto criado com sucesso",
	}, nil

}

File: frontend/src/application/usecase/CreatorProductUseCase.ts
import { Product } from '../../domain/entities/Product';
import type { HttpPostClient } from '../../domain/gateway/HttpPostClient';
import type { StorageClient } from '../../domain/gateway/StorageClient';
import { Name } from '../../domain/valuesobject/Name';
import { Price } from '../../domain/valuesobject/Price';
import { Quantity } from '../../domain/valuesobject/Quantity';
import { TokenError } from '../../shared/erros/TokenError';
import type { RequestCreatorProductDTO } from '../dto/RequestCreatorProductDTO';
import type { ResponseCreatorProductDTO } from '../dto/ResponseCreatorProductDTO';

export class CreatorProductUseCase {
  public constructor(
    private httpClient: HttpPostClient,
    private storageClient: StorageClient,
    private path: string,
    private key: string
  ) {}

  public async create(payload: RequestCreatorProductDTO): Promise<ResponseCreatorProductDTO> {
    try {
      const name: Name = Name.create(payload.name);
      const price: Price = Price.create(payload.price);
      const quantity: Quantity = Quantity.create(payload.quantity);
      const product: Product = Product.create(price, quantity, name);
      const { token } = this.storageClient.get<{ token: string }>(this.key);
      const registerResponse = await this.httpClient.post<ResponseCreatorProductDTO>(this.path, product.getProductForRegister(), {
        token,
      });
      return registerResponse;
    } catch (error) {
      error instanceof TokenError && this.storageClient.delete(this.key);
      throw error;
    }
  }
}
